pkg/controllers/observability/tempo_components.go

```diff
@@ -0,0 +1,300 @@
+		}
+
+		tempoSecret.Data = map[string][]byte{
+			"container":    []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.Azure.Container),
+			"account_name": []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.Azure.AccountName),
+			"account_key":  accountKeySecret.Data[instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.Azure.AccountKeySecret.Key],
+		}
+	} else if instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF != nil {
+		tempoSecret.Data = map[string][]byte{
+			"container":    []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF.Container),
+			"account_name": []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF.AccountName),
+			"client_id":    []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF.ClientID),
+			"tenant_id":    []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF.TenantID),
+		}
+		if instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF.Audience != "" {
+			tempoSecret.Data["audience"] = []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.AzureWIF.Audience)
+		}
+	} else if instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCS != nil {
+		keyJSONSecret := &corev1.Secret{}
+		err := k8sClient.Get(ctx, client.ObjectKey{
+			Namespace: instance.Namespace,
+			Name:      instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCS.KeyJSONSecret.Name,
+		}, keyJSONSecret)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get GCS keyJSON secret %s: %w", instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCS.KeyJSONSecret.Name, err)
+		}
+
+		tempoSecret.Data = map[string][]byte{
+			"bucketname": []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCS.Bucket),
+			"key.json":   keyJSONSecret.Data[instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCS.KeyJSONSecret.Key],
+		}
+	} else if instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCSSTSSpec != nil {
+		keyJSONSecret := &corev1.Secret{}
+		err := k8sClient.Get(ctx, client.ObjectKey{
+			Namespace: instance.Namespace,
+			Name:      instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCSSTSSpec.KeyJSONSecret.Name,
+		}, keyJSONSecret)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get GCSSTS keyJSON secret %s: %w", instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCSSTSSpec.KeyJSONSecret.Name, err)
+		}
+
+		tempoSecret.Data = map[string][]byte{
+			"bucketname": []byte(instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCSSTSSpec.Bucket),
+			"key.json":   keyJSONSecret.Data[instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCSSTSSpec.KeyJSONSecret.Key],
+		}
+		if instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec.GCSSTSSpec.Audience != "" {
```

Comment:
nit: Perhaps it would be clearer to extract `instance.Spec.Capabilities.Tracing.Storage.ObjectStorageSpec` into a new variable?

Reply:
+1, done